async_generator/_impl.py

```diff
@@ -1,94 +1,134 @@
+# to be able to send values to the context in which the generator is
+# being iterated, and it's awfully convenient to be able to do that
+# using 'yield' too. To distinguish between these two streams of
+# yielded objects, the traps intended for the event loop are yielded
+# as-is, and the values intended for the context that's iterating the
+# generator are wrapped in some wrapper object (YieldWrapper here, or
+# an internal Python type called AsyncGenWrappedValue in the native
+# async generator implementation) before being yielded.
+# The __anext__(), asend(), and athrow() methods of an async generator
+# iterate the underlying async function until a wrapped value is received,
+# and any unwrapped values are passed through to the event loop.
+
+# These functions are syntactically valid only on 3.6+, so we conditionally
+# exec() the code defining them.
+_native_asyncgen_helpers = """
+async def _wrapper():
+    holder = [None]
+    while True:
+        # The simpler "value = None; while True: value = yield value"
+        # would hold a reference to the most recently wrapped value
+        # after it has been yielded out (until the next value to wrap
+        # comes in), so we use a one-element list instead.
+        holder.append((yield holder.pop()))
+_wrapper = _wrapper()
+
+async def _unwrapper():
+    @coroutine
+    def inner():
+        holder = [None]
+        while True:
+            holder.append((yield holder.pop()))
+    await inner()
+    yield None
+_unwrapper = _unwrapper()
+"""
+
+if sys.implementation.name == "cpython" and sys.version_info >= (3, 6):
+    # On 3.6, with native async generators, we want to use the same
+    # wrapper type that native generators use. This lets @async_generators
+    # yield_from_ native async generators and vice versa.
+
+    import ctypes
+    from types import AsyncGeneratorType, GeneratorType
+    exec(_native_asyncgen_helpers)
+
+    # Transmute _wrapper to a regular generator object by modifying the
+    # ob_type field. The code object inside _wrapper will still think it's
+    # associated with an async generator, so it will yield out
+    # AsyncGenWrappedValues when it encounters a 'yield' statement;
+    # but the generator object will think it's a normal non-async
+    # generator, so it won't unwrap them. This way, we can obtain
+    # AsyncGenWrappedValues as normal manipulable Python objects.
+    #
+    # This sort of object type transmutation is categorically a Sketchy
+    # Thing To Do, because the functions associated with the new type
+    # (including tp_dealloc and so forth) will be operating on a
+    # structure whose in-memory layout matches that of the old type.
+    # In this case, it's OK, because async generator objects are just
+    # generator objects plus a few extra fields at the end; and these
+    # fields are two integers and a NULL-until-first-iteration object
+    # pointer, so they don't hold any resources that need to be cleaned up.
```

Comment:
Why is the `NULL-until-first-iteration object pointer` safe? I guess it's because we transmute the object before the first iteration, so the pointer remains NULL forever, and thus never needs to be DECREFed? This is probably worth noting explicitly.

Reply:
Yep, that's it exactly. I expanded the comment to clarify.